datastore.go

```diff
@@ -174,6 +174,13 @@ func (a *accessor) queryOrig(q dsq.Query) (dsq.Results, error) {
 		qr = dsq.NaiveFilter(qr, f)
 	}
 	for _, o := range q.Orders {
+		// leveldb iterators return entries sorted lexicographically by key, so we can bypass the naive ordering
+		// for OrderByKey. Since OrderByKey.Sort() has as struct receiver, the value stored in q.Orders can be a
+		// pointer or a struct, so we need to check for both for correctness.
+		switch o.(type) {
+		case dsq.OrderByKey, *dsq.OrderByKey:
+			continue
+		}
 		qr = dsq.NaiveOrder(qr, o)
```

Comment:
I believe we need `OrderByKey` to *override* other orders (this currently just skips these). Given that keys are unique and have a total order, I believe we should:

1. Start at the end.
2. Work backwards until we find the *last* `OrderByKey`.
3. Apply any remaining orders.

Reply:
We could, in theory, optimize `OrderByKeyDescending` as well by taking advantage of the same total ordering. That is, stop at the last `OrderByKeyDescending` (if we reach one before we reach an `OrderByKey`). Unlike `OrderByKey`, we'll have to actually *apply* the `OrderByKeyDescending` order.